examples/syscall-ftrace.c

```diff
@@ -0,0 +1,220 @@
+module_param(uid, int, 0644);
+
+/**
+ * This is a housekeeping structure that saves all information
+ * needed for hooking. Usage with `PREPARE_HOOK` is recommended.
+ * 
+ * Example:
+ * static ftrace_hook_t sys_clone_hook = 
+ *      PREPARE_HOOK(__NR_clone, my_sys_clone, &orig_sys_clone)
+ */
+typedef struct ftrace_hook {
+    unsigned long nr; // syscall number from unistd.h
+    void *new; // hook function
+    void *orig; // original function
+
+    unsigned long address; // address to the original function
+    struct ftrace_ops ops; // ftrace structure
+} ftrace_hook_t;
+
+#define PREPARE_HOOK(_nr, _hook, _orig)                                        \
+    {                                                                          \
+        .nr = (_nr), .new = (_hook), .orig = (_orig)                           \
+    }
+
+static unsigned long **sys_call_table;
+
+/**
+ * For the sake of simplicity, only the kprobe method is included.
+ * If you want to know more about different methods to get
+ * kallsyms_lookup_name, see syscall.c.
+ */
+static int resolve_address(ftrace_hook_t *hook)
+{
+    static struct kprobe kp = { .symbol_name = "kallsyms_lookup_name" };
+    unsigned long (*kallsyms_lookup_name)(const char *name);
+    register_kprobe(&kp);
+    kallsyms_lookup_name = (unsigned long (*)(const char *))kp.addr;
+    unregister_kprobe(&kp);
+
+    if (!kallsyms_lookup_name) {
+        pr_err("kallsyms_lookup_name is not found!\n");
+        return -1;
+    }
+    pr_info("kallsyms_lookup_name is found at 0x%lx\n",
+            (unsigned long)kallsyms_lookup_name);
+
+    sys_call_table = (unsigned long **)kallsyms_lookup_name("sys_call_table");
+    if (!sys_call_table) {
+        pr_err("sys_call_table is not found!\n");
+        return -1;
+    }
+    pr_info("sys_call_table is found at 0x%lx\n",
+            (unsigned long)sys_call_table);
+
+    hook->address = (unsigned long)sys_call_table[hook->nr];
+    *((unsigned long *)hook->orig) = hook->address;
+    return 0;
+}
+
+/**
+ * This is where the magic happens.
+ * 
+ * We check whether this function is called by the kernel or this module
+ * by checking whether parent_ip is within this module.
+ * 
+ * During the first call, parent_ip points to somewhere in the kernel
+ * that's not in this module,
+ * while the second call is in this module
+ * since it's called from our_sys_openat.
+ * 
+ * If it is the first call, we modify ip to be our_sys_openat,
+ * which will pass control to it after ftrace is done.
+ */
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 11, 0)
+static void notrace ftrace_thunk(unsigned long ip, unsigned long parent_ip,
+                                 struct ftrace_ops *ops,
+                                 struct ftrace_regs *fregs)
+{
+    ftrace_hook_t *hook = container_of(ops, ftrace_hook_t, ops);
+    if (!within_module(parent_ip, THIS_MODULE))
+        fregs->regs.ip = (unsigned long)hook->new;
+}
+
+#else /** Version < v5.11 */
```

Comment:
You used two different comment styles, `/** comment */` and `/* comment */`.
Please make sure they are consistent.